contrib/babelfishpg_tsql/src/pl_comp.c

```diff
@@ -1642,6 +1667,200 @@ pltsql_param_ref(ParseState *pstate, ParamRef *pref)
+			return NULL;	/* if not a valid column ref */
+		}
+		PG_END_TRY();
+		
+		is_geospatial_function = false;	/* resetting the global variable */
+		dataTypeId = ((Var *)col)->vartype;
+		/* Modifying colref to funcref since a valid Geospatial function call is identified */
+		funcid = LookupFuncName(list_make2(makeString("sys"), makeString(name)), 1, &dataTypeId, false);
+		funcexpr = makeFuncExpr(funcid, get_func_rettype(funcid), list_make1(col), 0, 0, COERCE_EXPLICIT_CALL);
+		funcexpr->location = cref->location;
+
+		return (Node *) funcexpr;
+	}
+
+	is_geospatial_function = false;	/* resetting the global variable */
+
+	return NULL;
+}
+
+/*
+ * This function identifies and modifies geospatial function call for
+ * STDistance(), STAsText(), STAsBinary() functions
+ */
+static List *resolve_geospatial_func_ref(ParseState *pstate, FuncCall *fn, List *fargs)
+{
+	Node *fname;	/* represents the geospatial function name */
+
+	if(list_length(fn->funcname) > 1 && list_length(fn->args) <= 1 
+		&& fn->agg_order == NULL && fn->agg_filter == NULL 
+		&& !fn->agg_within_group && !fn->agg_star 
+		&& !fn->agg_distinct && !fn->func_variadic)
+	{
+		Node *col;
+		bool flag = true;
+		fname = (Node *) llast(fn->funcname);
+		/* 
+		 * checking if last field matches any of the geospatial functions - stdistance, stastext, stasbinary
+		 * and also checking if is_geospatial_function variable is false to avoid entering the loop if it is not a function
+		 */
+		if (IsA(fname, String) && !is_geospatial_function &&
+				((pg_strcasecmp(strVal(fname), "stdistance") == 0 && list_length(fn->args) == 1) ||
+				(pg_strcasecmp(strVal(fname), "stastext") == 0 && list_length(fn->args) == 0) || 
+				(pg_strcasecmp(strVal(fname), "stasbinary") == 0 && list_length(fn->args) == 0)))
+		{
+			ColumnRef *cref = makeNode(ColumnRef);		/* new colref to avoid making changes to existing structure */
+			cref->fields = list_copy(fn->funcname);
+			cref->fields = list_delete_last(cref->fields);	/* deletes the function name and check if the remaining colref is a valid column reference */
+
+			/* if length is 1 then colref would have been modified by rewrite_plain_name function */
+			if(list_length(cref->fields) == 1)
+			{
+				Node       *schema = (Node *) linitial(cref->fields);
+				char       *cur_db = get_cur_db_name();
+				String     *new_schema;
+
+				if (!is_shared_schema(strVal(schema)))
+				{
+					new_schema = makeString(extract_logical_schema_name(cur_db, strVal(schema), get_migration_mode()));
+					cref->fields = list_delete_first(cref->fields);
+					cref->fields = lcons(new_schema, cref->fields);
+				}
+			}
+			is_geospatial_function = true;
+			PG_TRY();
+			{
+				col = transformExpr(pstate, (Node *) cref, pstate->p_expr_kind);	/* Recursive call to check for valid column reference */
+			}
+			PG_CATCH();
+			{
+				flag = false;	/* if not a valid column ref. Eg: table.col.stx (where col is not a coulmn of table) */
```

Comment:
Why can't we directly set is_geospatial_function=false? I don't see a need for this additional flag.

Reply:
Right, Fixed it